File: web/src/hive_tip.js
import {
  CELL_OPEN_D,
  CELL_OPEN_H,
  CELL_OPEN_W,
  CELL_SHELLS,
  EXTRA_NECTAR_PARK_Z,
  EXTRA_NECTAR_SPAWN,
  HIVE_PIVOT_Z,
  HIVE_TIP_JOINT_SIGN,
  HIVE_TIP_TRAVEL_DEG,
  HIVE_UPWARD_CELL,
  NECTAR_EXTRA_POOL,
  NECTAR_MAX_PER_ALLIANCE,
  NECTAR_R,
  NECTAR_STAGED_PER_ALLIANCE,
  TILE_THICKNESS,
  TIP_NECTAR_REQUIRED,
  TIP_POINTS,
  TIP_POLLEN_ALONE,
  TIP_POLLEN_WITH_NECTAR,
} from './constants.js';

function quatMul(q1, q2) {
  const [w1, x1, y1, z1] = q1;
  const [w2, x2, y2, z2] = q2;
  return [
    w1 * w2 - x1 * x2 - y1 * y2 - z1 * z2,
    w1 * x2 + x1 * w2 + y1 * z2 - z1 * y2,
    w1 * y2 - x1 * z2 + y1 * w2 + z1 * x2,
    w1 * z2 + x1 * y2 - y1 * x2 + z1 * w2,
  ];
}

function quatRx(rad) {
  const half = 0.5 * rad;
  return [Math.cos(half), Math.sin(half), 0, 0];
}

function rotXMat(rad) {
  const c = Math.cos(rad);
  const s = Math.sin(rad);
  return [
    [1, 0, 0],
    [0, c, -s],
    [0, s, c],
  ];
}

function matMulVec(M, v) {
  return [
    M[0][0] * v[0] + M[0][1] * v[1] + M[0][2] * v[2],
    M[1][0] * v[0] + M[1][1] * v[1] + M[1][2] * v[2],
    M[2][0] * v[0] + M[2][1] * v[1] + M[2][2] * v[2],
  ];
}

/**
 * Bi-stable HIVE tip both ways; each tip spawns +1 nectar up to 8 per alliance.
 * Tip animates about the pivot; free pieces inside either CELL are carried
 * rigidly so they stay captured (instant qpos teleport would spill out the back).
 */
export class HiveTipController {
  static COOLDOWN_S = 1.25;
  static ANIM_S = 0.4;
  static SETTLE_S = 0.2;

  constructor(mujoco, model, data) {
    this.mujoco = mujoco;
    this.model = model;
    this.data = data;
    this.tipState = { red: 0, blue: 0 };
    this.extrasReleased = { red: 0, blue: 0 };
    this.tipCount = { red: 0, blue: 0 };
    this.score = { red: 0, blue: 0 };
    this.cooldownUntil = { red: 0, blue: 0 };
    this.lastEvent = '';
    this._tipAnim = { red: null, blue: null };
    this.pivot = [0, 0, TILE_THICKNESS + HIVE_PIVOT_Z];

    const BODY = mujoco.mjtObj.mjOBJ_BODY.value;
    const JOINT = mujoco.mjtObj.mjOBJ_JOINT.value;

    this.jointQadr = {};
    this.jointDadr = {};
    this.shellBody = {};
    this.pollenIds = [];
    this.nectarIds = { red: [], blue: [] };
    this.extraIds = { red: [], blue: [] };
    this.extraQadr = { red: [], blue: [] };
    this.extraVadr = { red: [], blue: [] };
    this.bodyQadr = {};
    this.bodyVadr = {};

    for (const color of ['red', 'blue']) {
      const jid = mujoco.mj_name2id(model, JOINT, `hive_${color}_tip`);
      if (jid >= 0) {
        this.jointQadr[color] = model.jnt_qposadr[jid];
        this.jointDadr[color] = model.jnt_dofadr[jid];
      }
    }
    for (const key of Object.keys(CELL_SHELLS)) {
      const bid = mujoco.mj_name2id(model, BODY, `${key}_shell`);
      if (bid >= 0) this.shellBody[key] = bid;
    }
    for (let i = 0; i < model.nbody; i++) {
      const name = mujoco.mj_id2name(model, BODY, i);
      if (!name) continue;
      if (name.startsWith('pollen_')) this.pollenIds.push(i);
      else if (name.startsWith('nectar_red_') && !name.includes('extra')) this.nectarIds.red.push(i);
      else if (name.startsWith('nectar_blue_') && !name.includes('extra')) this.nectarIds.blue.push(i);
      else if (name.startsWith('nectar_extra_red_')) this.extraIds.red.push(i);
      else if (name.startsWith('nectar_extra_blue_')) this.extraIds.blue.push(i);
    }
    for (const color of ['red', 'blue']) {
      this.extraIds[color].sort((a, b) => {
        const na = mujoco.mj_id2name(model, BODY, a) || '';
        const nb = mujoco.mj_id2name(model, BODY, b) || '';
        return na.localeCompare(nb);
      });
      this.extraQadr[color] = [];
      this.extraVadr[color] = [];
      for (const bid of this.extraIds[color]) {
        const jnt = model.body_jntadr[bid];
        this.extraQadr[color].push(model.jnt_qposadr[jnt]);
        this.extraVadr[color].push(model.jnt_dofadr[jnt]);
      }
    }
    for (const bid of [
      ...this.pollenIds,
      ...this.nectarIds.red,
      ...this.nectarIds.blue,
      ...this.extraIds.red,
      ...this.extraIds.blue,
    ]) {
      const jnt = model.body_jntadr[bid];
      if (jnt < 0) continue;
      this.bodyQadr[bid] = model.jnt_qposadr[jnt];
      this.bodyVadr[bid] = model.jnt_dofadr[jnt];
    }
  }

  reset() {
    this.tipState = { red: 0, blue: 0 };
    this.extrasReleased = { red: 0, blue: 0 };
    this.tipCount = { red: 0, blue: 0 };
    this.score = { red: 0, blue: 0 };
    this.cooldownUntil = { red: 0, blue: 0 };
    this.lastEvent = '';
    this._tipAnim = { red: null, blue: null };
    for (const color of ['red', 'blue']) {
      if (this.jointQadr[color] != null) {
        this.data.qpos[this.jointQadr[color]] = 0;
        this.data.qvel[this.jointDadr[color]] = 0;
      }
      for (let i = 0; i < this.extraQadr[color].length; i++) this._parkExtra(color, i);
    }
  }

  update() {
    this._holdTipJoints();
    this._holdParkedExtras();
    const t = this.data.time;
    for (const color of ['red', 'blue']) {
      if (this.jointQadr[color] == null) continue;
      if (this._tipAnim[color] != null) continue;
      if (t < this.cooldownUntil[color]) continue;
      const [nNectar, nPollen] = this.countInUpwardCell(color);
      if (
        (nNectar >= TIP_NECTAR_REQUIRED && nPollen >= TIP_POLLEN_WITH_NECTAR) ||
        nPollen >= TIP_POLLEN_ALONE
      ) {
        this._doTip(color, nNectar, nPollen);
      }
    }
  }

  _tipTargetRad(color) {
    const sign = HIVE_TIP_JOINT_SIGN[color];
    if (this.tipState[color] === 0) return 0;
    return (sign * HIVE_TIP_TRAVEL_DEG * Math.PI) / 180;
  }

  _holdTipJoints() {
    const t = this.data.time;
    for (const color of ['red', 'blue']) {
      if (this.jointQadr[color] == null) continue;
      const anim = this._tipAnim[color];
      if (anim == null) {
        this.data.qpos[this.jointQadr[color]] = this._tipTargetRad(color);
        this.data.qvel[this.jointDadr[color]] = 0;
        continue;
      }
      const alpha = anim.duration > 0 ? (t - anim.startTime) / anim.duration : 1;
      if (alpha >= 1) {
        this._setJointAndPlaceCarry(color, anim, anim.qTo);
        if (t >= anim.holdUntil) this._tipAnim[color] = null;
      } else {
        const s = alpha * alpha * (3 - 2 * alpha);
        const q = anim.qFrom + s * (anim.qTo - anim.qFrom);
        this._setJointAndPlaceCarry(color, anim, q);
      }
    }
  }

  _setJointAndPlaceCarry(color, anim, q) {
    this.data.qpos[this.jointQadr[color]] = q;
    this.data.qvel[this.jointDadr[color]] = 0;
    const dq = q - anim.qFrom;
    const R = rotXMat(dq);
    const qRx = quatRx(dq);
    for (const [bidStr, snap] of Object.entries(anim.carry)) {
      const bid = Number(bidStr);
      const adr = this.bodyQadr[bid];
      const vadr = this.bodyVadr[bid];
      if (adr == null || vadr == null) continue;
      const pos = matMulVec(R, snap.rel);
      this.data.qpos[adr] = this.pivot[0] + pos[0];
      this.data.qpos[adr + 1] = this.pivot[1] + pos[1];
      this.data.qpos[adr + 2] = this.pivot[2] + pos[2];
      let quat = quatMul(qRx, snap.quat);
      const n = Math.hypot(quat[0], quat[1], quat[2], quat[3]);
      if (n > 1e-12) quat = quat.map((c) => c / n);
      this.data.qpos[adr + 3] = quat[0];
      this.data.qpos[adr + 4] = quat[1];
      this.data.qpos[adr + 5] = quat[2];
      this.data.qpos[adr + 6] = quat[3];
      for (let k = 0; k < 6; k++) this.data.qvel[vadr + k] = 0;
    }
  }

  _holdParkedExtras() {
    for (const color of ['red', 'blue']) {
      const nDone = this.extrasReleased[color];
      for (let i = nDone; i < this.extraQadr[color].length; i++) this._parkExtra(color, i);
    }
  }

  _hiveCellKeys(color) {
    return Object.keys(CELL_SHELLS).filter((k) => k.startsWith(`${color}_`));
  }

  _ballsInHiveCells(color) {
    const keys = this._hiveCellKeys(color);
    const candidates = [
      ...this.nectarIds[color],
      ...this.pollenIds,
      ...this.extraIds[color].slice(0, this.extrasReleased[color]),
    ];
    const out = [];
    const seen = new Set();
    for (const bid of candidates) {
      if (seen.has(bid)) continue;
      if (keys.some((key) => this._inCell(bid, key, -0.01))) {
        out.push(bid);
        seen.add(bid);
      }
    }
    return out;
  }

  countInUpwardCell(color) {
    const cellKey = HIVE_UPWARD_CELL[color][this.tipState[color]];
    let nNectar = 0;
    for (const bid of this.nectarIds[color]) {
      if (this._inCell(bid, cellKey)) nNectar++;
    }
    let nPollen = 0;
    for (const bid of this.pollenIds) {
      if (this._inCell(bid, cellKey)) nPollen++;
    }
    return [nNectar, nPollen];
  }

  _inCell(bid, cellKey, margin = 0.02) {
    const shellId = this.shellBody[cellKey];
    if (shellId == null) return false;
    const sOpen = CELL_SHELLS[cellKey].open_sign;
    const ox = this.data.xpos[shellId * 3];
    const oy = this.data.xpos[shellId * 3 + 1];
    const oz = this.data.xpos[shellId * 3 + 2];
    const b = shellId * 9;
    const R = this.data.xmat;
    const px = this.data.xpos[bid * 3] - ox;
    const py = this.data.xpos[bid * 3 + 1] - oy;
    const pz = this.data.xpos[bid * 3 + 2] - oz;
    const lx = R[b] * px + R[b + 3] * py + R[b + 6] * pz;
    const ly = R[b + 1] * px + R[b + 4] * py + R[b + 7] * pz;
    const lz = R[b + 2] * px + R[b + 5] * py + R[b + 8] * pz;
    const w = CELL_OPEN_W / 2 - margin;
    const h = CELL_OPEN_H / 2 - margin;
    const d = CELL_OPEN_D / 2;
    if (Math.abs(lx) > w || Math.abs(lz) > h) return false;
    const yOpen = sOpen * d;
    const yBack = -sOpen * d;
    const yLo = Math.min(yBack, yOpen) + margin;
    const yHi = Math.max(yBack, yOpen) - margin;
    return ly >= yLo && ly <= yHi;
  }

  _doTip(color, nNectar, nPollen) {
    const prev = this.tipState[color];
    const qFrom = this.data.qpos[this.jointQadr[color]];
    this.tipState[color] = 1 - prev;
    const qTo = this._tipTargetRad(color);
    const carry = {};
    for (const bid of this._ballsInHiveCells(color)) {
      const adr = this.bodyQadr[bid];
      if (adr == null) continue;
      carry[bid] = {
        rel: [
          this.data.qpos[adr] - this.pivot[0],
          this.data.qpos[adr + 1] - this.pivot[1],
          this.data.qpos[adr + 2] - this.pivot[2],
        ],
        quat: [
          this.data.qpos[adr + 3],
          this.data.qpos[adr + 4],
          this.data.qpos[adr + 5],
          this.data.qpos[adr + 6],
        ],
      };
    }
    const t = this.data.time;
    this._tipAnim[color] = {
      startTime: t,
      duration: HiveTipController.ANIM_S,
      qFrom,
      qTo,
      carry,
      holdUntil: t + HiveTipController.ANIM_S + HiveTipController.SETTLE_S,
    };
    this.cooldownUntil[color] =
      t + HiveTipController.COOLDOWN_S + HiveTipController.ANIM_S + HiveTipController.SETTLE_S;
    this.tipCount[color] += 1;
    this.score[color] += TIP_POINTS;
    const spawned = this._releaseExtra(color);
    const direction = this.tipState[color] === 1 ? 'TIP' : 'TIP BACK';
    const total = NECTAR_STAGED_PER_ALLIANCE + this.extrasReleased[color];
    const extraMsg = spawned
      ? `; +1 ${color} nectar (${total}/${NECTAR_MAX_PER_ALLIANCE})`
      : `; nectar cap ${NECTAR_MAX_PER_ALLIANCE}`;
    this.lastEvent = `HIVE ${color.toUpperCase()} ${direction} +${TIP_POINTS} (score ${this.score[color]}) — nectar=${nNectar} pollen=${nPollen}${extraMsg}`;
    this._holdTipJoints();
    this.mujoco.mj_forward(this.model, this.data);
  }

  _parkExtra(color, index) {
    const adr = this.extraQadr[color][index];
    const x = (color === 'blue' ? 0.15 : -0.15) + index * 0.02;
    this.data.qpos[adr] = x;
    this.data.qpos[adr + 1] = 0;
    this.data.qpos[adr + 2] = EXTRA_NECTAR_PARK_Z;
    this.data.qpos[adr + 3] = 1;
    this.data.qpos[adr + 4] = 0;
    this.data.qpos[adr + 5] = 0;
    this.data.qpos[adr + 6] = 0;
    const vadr = this.extraVadr[color][index];
    for (let k = 0; k < 6; k++) this.data.qvel[vadr + k] = 0;
  }

  _releaseExtra(color) {
    const idx = this.extrasReleased[color];
    const pool = this.extraQadr[color];
    if (idx >= pool.length || idx >= NECTAR_EXTRA_POOL) return false;
    let [sx, sy, sz] = EXTRA_NECTAR_SPAWN[color];
    sx += (color === 'blue' ? 0.05 : -0.05) * (idx % 3);
    sy += 0.04 * (Math.floor(idx / 3) % 3);
    sz = Math.max(sz, TILE_THICKNESS + NECTAR_R + 0.002);
    const adr = pool[idx];
    this.data.qpos[adr] = sx;
    this.data.qpos[adr + 1] = sy;
    this.data.qpos[adr + 2] = sz;
    this.data.qpos[adr + 3] = 1;
    this.data.qpos[adr + 4] = 0;
    this.data.qpos[adr + 5] = 0;
    this.data.qpos[adr + 6] = 0;
    const vadr = this.extraVadr[color][idx];
    for (let k = 0; k < 6; k++) this.data.qvel[vadr + k] = 0;
    this.extrasReleased[color] = idx + 1;
    return true;
  }
}

File: ftc_sim/hive_tip.py
"""BIOBUZZ HIVE bi-stable tipping: detect fill threshold, flip pose, spawn extra nectar."""

from __future__ import annotations

from dataclasses import dataclass, field

import mujoco
import numpy as np

from . import constants as C


def _quat_mul(q1: np.ndarray, q2: np.ndarray) -> np.ndarray:
    """MuJoCo (w, x, y, z) product q1 ⊗ q2."""
    w1, x1, y1, z1 = q1
    w2, x2, y2, z2 = q2
    return np.array(
        [
            w1 * w2 - x1 * x2 - y1 * y2 - z1 * z2,
            w1 * x2 + x1 * w2 + y1 * z2 - z1 * y2,
            w1 * y2 - x1 * z2 + y1 * w2 + z1 * x2,
            w1 * z2 + x1 * y2 - y1 * x2 + z1 * w2,
        ],
        dtype=float,
    )


def _quat_rx(rad: float) -> np.ndarray:
    half = 0.5 * rad
    return np.array([np.cos(half), np.sin(half), 0.0, 0.0], dtype=float)


def _rot_x_mat(rad: float) -> np.ndarray:
    c, s = np.cos(rad), np.sin(rad)
    return np.array([[1.0, 0.0, 0.0], [0.0, c, -s], [0.0, s, c]], dtype=float)


@dataclass
class _TipAnim:
    start_time: float
    duration: float
    q_from: float
    q_to: float
    # bid -> (rel_pos_about_pivot at tip start, quat at tip start)
    carry: dict[int, tuple[np.ndarray, np.ndarray]]
    hold_until: float = 0.0  # keep absolute place until this sim time


@dataclass
class HiveTipState:
    # tip_state 0 = CAD start pose; 1 = other bi-stable pose
    tip_state: dict[str, int] = field(default_factory=lambda: {"red": 0, "blue": 0})
    # How many tip-bonus nectar already spawned per alliance (cap = NECTAR_EXTRA_POOL)
    extras_released: dict[str, int] = field(default_factory=lambda: {"red": 0, "blue": 0})
    tip_count: dict[str, int] = field(default_factory=lambda: {"red": 0, "blue": 0})
    score: dict[str, int] = field(default_factory=lambda: {"red": 0, "blue": 0})
    cooldown_until: dict[str, float] = field(default_factory=lambda: {"red": 0.0, "blue": 0.0})
    last_event: str = ""


class HiveTipController:
    """
    Bi-stable tip when the *currently upward* CELL holds:
      - ≥3 alliance NECTAR + ≥4 POLLEN, or
      - ≥8 POLLEN alone.
    Tips either direction: filling the other cell tips the hive back.
    Each tip: release one parked extra nectar (up to NECTAR_MAX_PER_ALLIANCE total).

    Tip animates about ±X through the HIVE pivot. Free nectar/pollen inside either
    CELL of that hive are carried with the same Rx so they stay captured (an
    instantaneous qpos teleport would leave them behind and spill out the back).
    """

    COOLDOWN_S = 1.25  # settle after a flip before counting again
    ANIM_S = 0.40  # gradual tip duration (seconds)
    SETTLE_S = 0.20  # keep carrying at final pose so contacts form

    def __init__(self, model: mujoco.MjModel, data: mujoco.MjData) -> None:
        self.model = model
        self.data = data
        self.state = HiveTipState()
        self._joint_qadr: dict[str, int] = {}
        self._joint_dadr: dict[str, int] = {}
        self._shell_body: dict[str, int] = {}
        self._pollen_ids: list[int] = []
        self._nectar_ids: dict[str, list[int]] = {"red": [], "blue": []}
        # extras[color] = list of body ids in release order
        self._extra_ids: dict[str, list[int]] = {"red": [], "blue": []}
        self._extra_qadr: dict[str, list[int]] = {"red": [], "blue": []}
        self._extra_vadr: dict[str, list[int]] = {"red": [], "blue": []}
        self._body_qadr: dict[int, int] = {}
        self._body_vadr: dict[int, int] = {}
        self._tip_anim: dict[str, _TipAnim | None] = {"red": None, "blue": None}
        self._pivot = np.array([0.0, 0.0, C.TILE_THICKNESS + C.HIVE_PIVOT_Z], dtype=float)
        self._resolve_ids()

    def _resolve_ids(self) -> None:
        m = self.model
        for color, jname in (("red", "hive_red_tip"), ("blue", "hive_blue_tip")):
            jid = mujoco.mj_name2id(m, mujoco.mjtObj.mjOBJ_JOINT, jname)
            if jid >= 0:
                self._joint_qadr[color] = int(m.jnt_qposadr[jid])
                self._joint_dadr[color] = int(m.jnt_dofadr[jid])
        for key in C.CELL_SHELLS:
            bid = mujoco.mj_name2id(m, mujoco.mjtObj.mjOBJ_BODY, f"{key}_shell")
            if bid >= 0:
                self._shell_body[key] = bid
        for i in range(m.nbody):
            name = mujoco.mj_id2name(m, mujoco.mjtObj.mjOBJ_BODY, i)
            if not name:
                continue
            if name.startswith("pollen_"):
                self._pollen_ids.append(i)
            elif name.startswith("nectar_red_") and "extra" not in name:
                self._nectar_ids["red"].append(i)
            elif name.startswith("nectar_blue_") and "extra" not in name:
                self._nectar_ids["blue"].append(i)
            elif name.startswith("nectar_extra_red_"):
                self._extra_ids["red"].append(i)
            elif name.startswith("nectar_extra_blue_"):
                self._extra_ids["blue"].append(i)
        for color in ("red", "blue"):
            bodies = self._extra_ids[color]
            bodies.sort(
                key=lambda bid: mujoco.mj_id2name(m, mujoco.mjtObj.mjOBJ_BODY, bid) or ""
            )
            self._extra_ids[color] = bodies
            self._extra_qadr[color] = []
            self._extra_vadr[color] = []
            for bid in bodies:
                jnt = m.body_jntadr[bid]
                self._extra_qadr[color].append(int(m.jnt_qposadr[jnt]))
                self._extra_vadr[color].append(int(m.jnt_dofadr[jnt]))
        for bid in (
            self._pollen_ids
            + self._nectar_ids["red"]
            + self._nectar_ids["blue"]
            + self._extra_ids["red"]
            + self._extra_ids["blue"]
        ):
            jnt = m.body_jntadr[bid]
            if jnt < 0:
                continue
            self._body_qadr[bid] = int(m.jnt_qposadr[jnt])
            self._body_vadr[bid] = int(m.jnt_dofadr[jnt])

    def reset(self) -> None:
        self.state = HiveTipState()
        self._tip_anim = {"red": None, "blue": None}
        for color in ("red", "blue"):
            if color in self._joint_qadr:
                self.data.qpos[self._joint_qadr[color]] = 0.0
                self.data.qvel[self._joint_dadr[color]] = 0.0
            for i in range(len(self._extra_qadr.get(color, []))):
                self._park_extra(color, i)

    def update(self) -> None:
        self._hold_tip_joints()
        self._hold_parked_extras()
        t = float(self.data.time)
        for color in ("red", "blue"):
            if color not in self._joint_qadr:
                continue
            if self._tip_anim[color] is not None:
                continue
            if t < self.state.cooldown_until[color]:
                continue
            n_nectar, n_pollen = self.count_in_upward_cell(color)
            if (n_nectar >= C.TIP_NECTAR_REQUIRED and n_pollen >= C.TIP_POLLEN_WITH_NECTAR) or (
                n_pollen >= C.TIP_POLLEN_ALONE
            ):
                self._do_tip(color, n_nectar, n_pollen)

    def _tip_target_rad(self, color: str) -> float:
        sign = C.HIVE_TIP_JOINT_SIGN[f"hive_{color}"]
        if self.state.tip_state[color] == 0:
            return 0.0
        return float(np.radians(sign * C.HIVE_TIP_TRAVEL_DEG))

    def _hold_tip_joints(self) -> None:
        t = float(self.data.time)
        for color in ("red", "blue"):
            if color not in self._joint_qadr:
                continue
            anim = self._tip_anim[color]
            if anim is None:
                self.data.qpos[self._joint_qadr[color]] = self._tip_target_rad(color)
                self.data.qvel[self._joint_dadr[color]] = 0.0
                continue
            alpha = (t - anim.start_time) / anim.duration if anim.duration > 0 else 1.0
            if alpha >= 1.0:
                self._set_joint_and_place_carry(color, anim, anim.q_to)
                if t >= anim.hold_until:
                    self._tip_anim[color] = None
            else:
                s = alpha * alpha * (3.0 - 2.0 * alpha)
                q = anim.q_from + s * (anim.q_to - anim.q_from)
                self._set_joint_and_place_carry(color, anim, q)

    def _set_joint_and_place_carry(self, color: str, anim: _TipAnim, q: float) -> None:
        """Hold tip joint at q and rigidly place carried balls from tip-start snapshot."""
        self.data.qpos[self._joint_qadr[color]] = q
        self.data.qvel[self._joint_dadr[color]] = 0.0
        # Delta from CAD tip-start (joint=0) — snapshot rel was taken at q_from
        dq_from_start = float(q - anim.q_from)
        R = _rot_x_mat(dq_from_start)
        q_rx = _quat_rx(dq_from_start)
        for bid, (rel0, quat0) in anim.carry.items():
            adr = self._body_qadr.get(bid)
            vadr = self._body_vadr.get(bid)
            if adr is None or vadr is None:
                continue
            pos = self._pivot + (R @ rel0)
            quat = _quat_mul(q_rx, quat0)
            n = np.linalg.norm(quat)
            if n > 1e-12:
                quat = quat / n
            self.data.qpos[adr : adr + 3] = pos
            self.data.qpos[adr + 3 : adr + 7] = quat
            self.data.qvel[vadr : vadr + 6] = 0.0

    def _hold_parked_extras(self) -> None:
        for color in ("red", "blue"):
            n_done = self.state.extras_released[color]
            for i in range(n_done, len(self._extra_qadr.get(color, []))):
                self._park_extra(color, i)

    def _hive_cell_keys(self, color: str) -> list[str]:
        return [k for k, spec in C.CELL_SHELLS.items() if spec.get("color") == color]

    def _balls_in_hive_cells(self, color: str) -> list[int]:
        keys = self._hive_cell_keys(color)
        candidates = list(self._nectar_ids[color]) + list(self._pollen_ids)
        n_done = self.state.extras_released[color]
        candidates.extend(self._extra_ids[color][:n_done])
        out: list[int] = []
        seen: set[int] = set()
        for bid in candidates:
            if bid in seen:
                continue
            if any(self._in_cell(bid, key, margin=-0.01) for key in keys):
                out.append(bid)
                seen.add(bid)
        return out

    def count_in_upward_cell(self, color: str) -> tuple[int, int]:
        cell_key = C.HIVE_UPWARD_CELL[color][self.state.tip_state[color]]
        n_nectar = sum(1 for bid in self._nectar_ids[color] if self._in_cell(bid, cell_key))
        n_pollen = sum(1 for bid in self._pollen_ids if self._in_cell(bid, cell_key))
        return n_nectar, n_pollen

    def _in_cell(self, bid: int, cell_key: str, margin: float = 0.02) -> bool:
        shell_id = self._shell_body.get(cell_key)
        if shell_id is None:
            return False
        spec = C.CELL_SHELLS[cell_key]
        s_open = float(spec.get("open_sign", 1.0))
        origin = np.array(self.data.xpos[shell_id], dtype=float)
        R = self.data.xmat[shell_id].reshape(3, 3)
        pos = np.array(self.data.xpos[bid], dtype=float)
        local = R.T @ (pos - origin)
        w = C.CELL_OPEN_W / 2 - margin
        h = C.CELL_OPEN_H / 2 - margin
        d = C.CELL_OPEN_D / 2
        if abs(local[0]) > w or abs(local[2]) > h:
            return False
        y_open = s_open * d
        y_back = -s_open * d
        y_lo, y_hi = (min(y_back, y_open) + margin, max(y_back, y_open) - margin)
        return y_lo <= local[1] <= y_hi

    def _do_tip(self, color: str, n_nectar: int, n_pollen: int) -> None:
        prev = self.state.tip_state[color]
        q_from = float(self.data.qpos[self._joint_qadr[color]])
        self.state.tip_state[color] = 1 - prev
        q_to = self._tip_target_rad(color)
        carry: dict[int, tuple[np.ndarray, np.ndarray]] = {}
        for bid in self._balls_in_hive_cells(color):
            adr = self._body_qadr.get(bid)
            if adr is None:
                continue
            pos = np.array(self.data.qpos[adr : adr + 3], dtype=float)
            quat = np.array(self.data.qpos[adr + 3 : adr + 7], dtype=float)
            carry[bid] = (pos - self._pivot, quat.copy())
        t = float(self.data.time)
        self._tip_anim[color] = _TipAnim(
            start_time=t,
            duration=self.ANIM_S,
            q_from=q_from,
            q_to=q_to,
            carry=carry,
            hold_until=t + self.ANIM_S + self.SETTLE_S,
        )
        self.state.cooldown_until[color] = t + self.COOLDOWN_S + self.ANIM_S + self.SETTLE_S
        self.state.tip_count[color] += 1
        self.state.score[color] += C.TIP_POINTS
        spawned = self._release_extra_nectar(color)
        direction = "TIP" if self.state.tip_state[color] == 1 else "TIP BACK"
        total = C.NECTAR_STAGED_PER_ALLIANCE + self.state.extras_released[color]
        extra_msg = (
            f"; +1 {color} nectar (total {total}/{C.NECTAR_MAX_PER_ALLIANCE})"
            if spawned
            else f"; nectar cap {C.NECTAR_MAX_PER_ALLIANCE}"
        )
        self.state.last_event = (
            f"HIVE {color.upper()} {direction} +{C.TIP_POINTS} "
            f"(score {self.state.score[color]}) — nectar={n_nectar} pollen={n_pollen}{extra_msg}"
        )
        self._hold_tip_joints()
        mujoco.mj_forward(self.model, self.data)

    def _park_extra(self, color: str, index: int) -> None:
        adr = self._extra_qadr[color][index]
        x = (0.15 if color == "blue" else -0.15) + index * 0.02
        self.data.qpos[adr : adr + 3] = np.array([x, 0.0, C.EXTRA_NECTAR_PARK_Z])
        self.data.qpos[adr + 3 : adr + 7] = np.array([1.0, 0.0, 0.0, 0.0])
        vadr = self._extra_vadr[color][index]
        self.data.qvel[vadr : vadr + 6] = 0.0

    def _release_extra_nectar(self, color: str) -> bool:
        """Spawn one tip-bonus nectar if under the per-alliance cap. Returns True if spawned."""
        idx = self.state.extras_released[color]
        pool = self._extra_qadr.get(color, [])
        if idx >= len(pool) or idx >= C.NECTAR_EXTRA_POOL:
            return False
        sx, sy, sz = C.EXTRA_NECTAR_SPAWN[color]
        sx = sx + (0.05 if color == "blue" else -0.05) * (idx % 3)
        sy = sy + 0.04 * ((idx // 3) % 3)
        sz = max(sz, C.TILE_THICKNESS + C.NECTAR_R + 0.002)
        adr = pool[idx]
        self.data.qpos[adr : adr + 3] = np.array([sx, sy, sz])
        self.data.qpos[adr + 3 : adr + 7] = np.array([1.0, 0.0, 0.0, 0.0])
        vadr = self._extra_vadr[color][idx]
        self.data.qvel[vadr : vadr + 6] = 0.0
        self.state.extras_released[color] = idx + 1
        return True
